File: packages/core_dart/lib/src/conditional_style.dart
/// KRO-198 — matcher de "Estilo por valor" (ConditionalStyle): evalúa un caso
/// `{op, value}` contra el dato `raw`. Espejo PURO de `matchConditionalCase`
/// (`conditional-style.ts`). Ops: `eq` (def) · `neq` · `contains` · `gt`/`gte`/
/// `lt`/`lte` (numéricas) · `truthy`/`falsy` (ignoran value). Texto: trim +
/// case-insensitive.
///
/// KRO-198 — además del primitivo de matching, este módulo resuelve la APARIENCIA
/// condicional completa (`resolveConditionalStyling`/`matchedConditionalCase`/
/// `resolveConditionalAppearance`), espejo de `conditional-style.ts`. Lo usa el
/// render de slots/chips (el "estilo por valor": color por rareza/elemento…).
library;

import 'composition.dart' show ConditionalStyle, ConditionalStyleCase, SlotAppearance;

/// Parsea a número (acepta strings numéricos); null si no es número finito.
num? _asNumber(Object? v) {
  if (v is num) return v.isFinite ? v : null;
  if (v is String) {
    final t = v.trim();
    if (t.isEmpty) return null;
    return num.tryParse(t);
  }
  return null;
}

/// TS `!!raw && raw !== '0' && raw !== 'false'`.
bool _truthy(Object? raw) {
  if (raw == null) return false;
  if (raw is bool) return raw;
  if (raw is num) return raw != 0;
  final s = raw.toString();
  return s.isNotEmpty && s != '0' && s != 'false';
}

/// ¿El valor `raw` del dato cumple el caso `{op, value}`? Espejo 1:1 de
/// `matchConditionalCase`.
bool matchConditionalCase({String? op, String? value, required Object? raw}) {
  final o = op ?? 'eq';
  if (o == 'truthy') return _truthy(raw);
  if (o == 'falsy') return !_truthy(raw);

  if (o == 'gt' || o == 'gte' || o == 'lt' || o == 'lte') {
    final a = _asNumber(raw);
    final b = _asNumber(value);
    if (a == null || b == null) return false;
    return switch (o) {
      'gt' => a > b,
      'gte' => a >= b,
      'lt' => a < b,
      _ => a <= b, // lte
    };
  }

  final s = (raw == null ? '' : raw.toString()).trim().toLowerCase();
  final t = (value ?? '').trim().toLowerCase();
  if (o == 'contains') return t.isNotEmpty && s.contains(t);
  if (o == 'neq') return s != t;
  return s == t; // eq
}

/// KRO-198 — el CASO que matchea (no el merge): el 1º de `cond.cases` cuyo `{op,
/// value}` cumple el valor de `cond.fieldKey` en `item`. Para leer su `target` y
/// aplicar la apariencia a los chips correctos. NO contempla el `otherwise` (else)
/// — puro "primer caso que coincide". Espejo de `matchedConditionalCase`.
ConditionalStyleCase? matchedConditionalCase(
    ConditionalStyle? cond, Map<String, dynamic>? item) {
  if (cond == null || cond.fieldKey.isEmpty || cond.cases.isEmpty || item == null) {
    return null;
  }
  final raw = item[cond.fieldKey];
  for (final c in cond.cases) {
    if (matchConditionalCase(op: c.op, value: c.value, raw: raw)) return c;
  }
  return null;
}

/// KRO-198 — caso EFECTIVO CON la cláusula else: el 1º `case` que coincide o, si
/// ninguno, `cond.otherwise`; `null` si no hay ni match ni else. El else aplica SOLO
/// con el condicional configurado (fieldKey + cases) y evaluado contra datos (hay
/// item). Es el helper que usa el render (para que el else surta efecto). Espejo de
/// `resolveConditionalStyling`.
ConditionalStyleCase? resolveConditionalStyling(
    ConditionalStyle? cond, Map<String, dynamic>? item) {
  if (cond == null || cond.fieldKey.isEmpty || cond.cases.isEmpty || item == null) {
    return null;
  }
  return matchedConditionalCase(cond, item) ?? cond.otherwise;
}

/// KRO-198 — apariencia EFECTIVA tras el estilo condicional: la del caso efectivo
/// mergeada sobre `base` (el caso gana). Sin condicional/match/else → `base` intacta.
/// Espejo de `resolveConditionalAppearance` (para slots NO composables — sin chips).
SlotAppearance? resolveConditionalAppearance(
    ConditionalStyle? cond, SlotAppearance? base, Map<String, dynamic>? item) {
  final eff = resolveConditionalStyling(cond, item);
  if (eff == null) return base;
  final ap = eff.appearance;
  return ap == null ? base : ap.mergedOver(base);
}

File: packages/core/src/conditional-style.ts
/**
 * KRO-198 — Estilo condicional por valor: evalúa `SlotComposition.conditionalStyle`
 * contra el dato de una carta y devuelve la apariencia efectiva.
 *
 * Resuelve "color/estilo por rareza" (y similares) de forma DECLARATIVA, dentro
 * del propio modelo de appearance — sin depender del sistema overlay de
 * visual-effects. Puro: mismas entradas → misma salida. Flutter lo espeja.
 */
import type { ConditionalStyle, ConditionalStyleCase, SlotAppearance } from './types';

/** Parsea un valor a número (acepta strings numéricos); null si no es número. */
function asNumber(v: unknown): number | null {
  if (typeof v === 'number' && Number.isFinite(v)) return v;
  if (typeof v === 'string' && v.trim() !== '' && !Number.isNaN(Number(v))) return Number(v);
  return null;
}

/** ¿El valor `raw` del dato cumple este caso? Comparación de texto
 *  case-insensitive + trim; gt/gte/lt/lte numéricas; truthy/falsy ignoran value. */
export function matchConditionalCase(c: ConditionalStyleCase, raw: unknown): boolean {
  const op = c.op ?? 'eq';
  if (op === 'truthy') return !!raw && raw !== '0' && raw !== 'false';
  if (op === 'falsy')  return !raw || raw === '0' || raw === 'false';

  if (op === 'gt' || op === 'gte' || op === 'lt' || op === 'lte') {
    const a = asNumber(raw), b = asNumber(c.value);
    if (a == null || b == null) return false;
    return op === 'gt' ? a > b : op === 'gte' ? a >= b : op === 'lt' ? a < b : a <= b;
  }

  const s  = (raw == null ? '' : String(raw)).trim().toLowerCase();
  const t  = (c.value ?? '').trim().toLowerCase();
  if (op === 'contains') return t !== '' && s.includes(t);
  if (op === 'neq')      return s !== t;
  return s === t; // eq
}

/**
 * Apariencia EFECTIVA tras aplicar el estilo condicional: el primer caso que
 * matchea el valor de `cond.fieldKey` en `item` MERGE-a su appearance sobre la
 * base; si ninguno matchea pero hay cláusula `otherwise` (else), esa manda. Sin
 * condicional / sin match / sin else → devuelve la base intacta.
 */
export function resolveConditionalAppearance(
  cond: ConditionalStyle | undefined,
  base: SlotAppearance | undefined,
  item: Record<string, unknown> | undefined,
): SlotAppearance | undefined {
  const eff = resolveConditionalStyling(cond, item);
  if (!eff) return base;
  return { ...(base ?? {}), ...(eff.appearance ?? {}) };
}

/**
 * KRO-198 — devuelve el CASO que matchea (no solo el merge), para que el caller pueda
 * leer su `target` y aplicar la apariencia a los chip(s) correctos del slot componible
 * (ganando sobre su apariencia por-chip) en vez de a la base de toda la fila. Sin
 * condicional / sin match / sin item → undefined. Mismo orden de evaluación (1º que
 * matchea gana) que `resolveConditionalAppearance`. NO contempla el `otherwise` (else):
 * es puro "primer caso que coincide" → para el fallback usa `resolveConditionalStyling`.
 */
export function matchedConditionalCase(
  cond: ConditionalStyle | undefined,
  item: Record<string, unknown> | undefined,
): ConditionalStyleCase | undefined {
  if (!cond?.fieldKey || !cond.cases?.length || !item) return undefined;
  const raw = item[cond.fieldKey];
  return cond.cases.find(c => matchConditionalCase(c, raw));
}

/**
 * KRO-198 — caso EFECTIVO del estilo condicional, CON la cláusula else: el primer
 * `case` que coincide o, si ninguno, la cláusula `otherwise` (else); undefined si no
 * hay ni match ni else. El caller lee `.appearance` + `.target` IGUAL para casos y
 * para el else → el "aplica a"/scoping por-chip funciona idéntico en ambos. Este es
 * el helper que debe usar el render (resolveSlot) para que el else surta efecto.
 */
export function resolveConditionalStyling(
  cond: ConditionalStyle | undefined,
  item: Record<string, unknown> | undefined,
): ConditionalStyleCase | undefined {
  // El else aplica SOLO cuando el condicional está configurado (fieldKey + cases) y
  // se EVALUÓ contra datos (hay item) pero ningún caso coincidió. Sin configurar /
  // sin item → undefined (no filtra el else sin datos que evaluar).
  if (!cond?.fieldKey || !cond.cases?.length || !item) return undefined;
  return matchedConditionalCase(cond, item) ?? cond.otherwise;
}
